sam/io.go

```diff
@@ -40,17 +40,158 @@ func ReadToChan(filename string, data chan<- Sam, header chan<- Header) {
 	close(data)
 }
 
+// readSamToChanRecycle is similar to readSamToChan however it reuses Sam structs from receiveRecords avoiding
+// allocations for each new read as is necessary in readSamToChan.
+func readSamToChanRecycle(filename string, sendRecords chan<- *Sam, receiveRecords <-chan *Sam, header chan<- Header) {
+	var file *fileio.EasyReader
+	var curr *Sam
+	var done bool
+	var err error
+
+	file = fileio.EasyOpen(filename)
+
+	header <- ReadHeader(file)
+	close(header)
+
+	for curr = range receiveRecords {
+		done = readNextRecycle(file, curr)
+		if done {
+			break
+		}
+		sendRecords <- curr
+	}
+
+	err = file.Close()
+	exception.PanicOnErr(err)
+	close(sendRecords)
+}
+
+// readBamToChan streams the input Bam file to the input data and header channel
+// so that only a small portion of the file is kept in memory at a time.
+// the header channel will have a single send (the Header struct) then the
+// channel will be closed. The header can be retrieved by `header := <-headerChan`.
+// Records will continuously stream to the input data channel until the end of the
+// file is reached at which point the data channel will be closed.
+func readBamToChan(filename string, data chan<- Sam, header chan<- Header) {
+	var file *BamReader
+	var head Header
+	var err error
+
+	file, head = OpenBam(filename)
+
+	header <- head
+	close(header)
+
+	for {
+		var curr Sam
+		_, err = DecodeBam(file, &curr)
+		if err == io.EOF {
+			break
+		}
+		data <- curr
+	}
+
+	err = file.Close()
+	exception.PanicOnErr(err)
+	close(data)
+}
+
+// readBamToChanRecycle is similar to readBamToChan however it reuses Sam structs from receiveRecords avoiding
+// allocations for each new read as is necessary in readBamToChan.
+func readBamToChanRecycle(filename string, sendRecords chan<- *Sam, receiveRecords <-chan *Sam, header chan<- Header) {
+	var file *BamReader
+	var head Header
+	var curr *Sam
+	var err error
+
+	file, head = OpenBam(filename)
+
+	header <- head
+	close(header)
+
+	for curr = range receiveRecords {
+		_, err = DecodeBam(file, curr)
+		if err == io.EOF {
+			break
+		}
+		sendRecords <- curr
+	}
+
+	err = file.Close()
+	exception.PanicOnErr(err)
+	close(sendRecords)
+}
+
 // GoReadToChan streams the input file so that only a small portion
-// of the file is kept in memory at a time. This function wraps the
-// ReadToChan function to automatically handle channel creation,
-// goroutine spawning, and header retrieval.
+// of the file is kept in memory at a time. This function automatically
+// handles channel creation, goroutine spawning, and header retrieval.
+//
+// GoReadToChan will detect if the input file ends in ".bam" and
+// automatically switch to a bam parser.
 func GoReadToChan(filename string) (<-chan Sam, Header) {
 	data := make(chan Sam, 1000)
 	header := make(chan Header)
-	go ReadToChan(filename, data, header)
+
+	if strings.HasSuffix(filename, ".bam") {
+		go readBamToChan(filename, data, header)
+	} else {
```

Comment:
Does it make sense to have another if for .sam and then else is a fatal or something. That way we catch is someone puts into something that is neither bam or sam? There could be something upstream that already does this, and I'm just missing it.